src/Moq/Expressions/Visitors/ConstructorCallVisitor.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) 2007, Clarius Consulting, Manas Technology Solutions, InSTEDD.
+// All rights reserved. Licensed under the BSD 3-Clause License; see License.txt.
+
+using System;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+
+using Moq.Properties;
+
+namespace Moq.Expressions.Visitors
+{
+	internal class ConstructorCallVisitor : ExpressionVisitor
+	{
+		/// <summary>
+		/// Extracts the arguments from a lambda expression that calls a constructor.
+		/// </summary>
+		/// <param name="newExpression">The constructor expression.</param>
+		/// <returns>Extracted argument values.</returns>
+		public static object[] ExtractArgumentValues(LambdaExpression newExpression)
+		{
+			if (newExpression is null)
+			{
+				throw new ArgumentNullException(nameof(newExpression));
+			}
+
+			var visitor = new ConstructorCallVisitor();
+			visitor.Visit(newExpression);
+
+			if (visitor._constructor == null)
+			{
+				throw new NotSupportedException(Resources.NoConstructorCallFound);
+			}
+
+			return visitor._arguments;
+		}
+
+		private ConstructorInfo _constructor;
+		private object[] _arguments;
+
+		public override Expression Visit(Expression node)
+		{
+			switch (node)
+			{
+				case LambdaExpression _:
+				case NewExpression _:
+				case UnaryExpression unary when unary.NodeType == ExpressionType.Quote:
```

Comment:
Please rewrite the `switch` to match `node.NodeType` instead of `node`. That'll quite possibly be more efficient and it'll actually simplify the third `case` quite a bit.